src/backend/distributed/executor/partitioned_intermediate_results.c

```diff
@@ -501,22 +500,17 @@ PartitionedResultDestReceiverShutdown(DestReceiver *copyDest)
  * PartitionedResultDestReceiver.
  */
 static void
-PartitionedResultDestReceiverDestroy(DestReceiver *copyDest)
+PartitionedResultDestReceiverDestroy(DestReceiver *dest)
 {
-	PartitionedResultDestReceiver *partitionedDest =
-		(PartitionedResultDestReceiver *) copyDest;
-	int partitionCount = partitionedDest->partitionCount;
-	for (int partitionIndex = 0; partitionIndex < partitionCount; partitionIndex++)
+	PartitionedResultDestReceiver *self = (PartitionedResultDestReceiver *) dest;
+
+	for (int partitionIndex = 0; partitionIndex < self->partitionCount; partitionIndex++)
 	{
-		DestReceiver *partitionDest =
-			partitionedDest->partitionDestReceivers[partitionIndex];
+		/* TODO understand if we should also destroy non-started receivers */
```

Comment:
I'm inlined towards no. DestReceivers normally go through at least an rStartup -> rShutdown sequence before going through rDestroy.

Reply:
Makes sense, and I agree. However, how the FileDestReceiver's `rDestroy` is used could benefit from it always being called, although I think that use is wrong.

We are de-allocating all kinds of field manually, instead of relying on the `MemoryContext` is was created in to reclaim the memory. What is your take on that? Would it make sense to remove our `pfree`'s from the destroy? If it really makes sense to `pfree` the data it probably should be done not by the `DestReceiver` but instead by the controlling code.

Reply:
I don't think rDestroy actually gets called by the current caller so the implementation is somewhat arbitrary. Even if it did get called, it does not currently do anything particularly useful, because the statement ends immediately after we stop using the DestReceiver, which resets the memory context in which the DestReceiver was created. It think rDestroy exists mainly to do some micro-optimizations.

There are some useful comments on the contract of DestReceivers in src/include/tcop/dest.h 

Not strictly necessary for our purpose, but we should maybe ensure reusability:

>  * CreateDestReceiver returns a receiver object appropriate to the specified
>  * destination.  The executor, as well as utility statements that can return
>  * tuples, are passed the resulting DestReceiver* pointer.  Each executor run
>  * or utility execution calls the receiver's rStartup method, then the
>  * receiveSlot method (zero or more times), then the rShutdown method.
>  * The same receiver object may be re-used multiple times; eventually it is
>  * destroyed by calling its rDestroy method.

Reply:
if reusability is a thing, we should actually destroy irregardless of it being started or not.

If the `PartitionResultDestReceiver` is reused it might on a subsequent run not start the same receivers in lazy mode. Meaning receivers having been started on an earlier run but not on the last run will not be destroyed.

Reply:
but destroy should only be called after the last run (see comment above)

maybe we just ignore destroy? (can you check whether it's actually called?)

Reply:
Destroy is not called, neither hitting a breakpoint in my environment, nor coverage on the main branch reported in codecov: https://app.codecov.io/gh/citusdata/citus/blob/master/src/backend/distributed/executor/partitioned_intermediate_results.c

Reply:
Makes sense, rDestroy calls are always done separately outside of Portal calls, so we'd have to call it ourselves.

Reply:
as you might have seen, I actually added the `rDestroy` call to make it covered in tests as well.